File: leetcode1819.cpp
#include <bits/stdc++.h>
using namespace std;
class Solution {
public:
    int countDifferentSubsequenceGCDs(vector<int>& nums) {
        int ans = 0;
        int maxx = 0;
        std::for_each(nums.begin(), nums.end(), [&](const auto &item) {
            maxx = max(maxx,item);
        });
        vector<bool> vis(maxx + 1);
        for (auto i : nums) {
            vis[i] = true;
        }
        for (int i = 1;i <= maxx;i ++) {
            int gcd = 0;
            for (int j = i;j <= maxx;j += i) {
                if (vis[j]) {
                    if (gcd == 0){
                        gcd = j;
                    } else {
                        gcd = __gcd(gcd,j);
                    }
                    if (gcd == i) {
                        ans ++;
                        break;
                    }
                }
            }
        }
        return ans;
    }
};
int main() {
    vector<int> faq = {16,15};
    cout<<(new Solution)->countDifferentSubsequenceGCDs(faq);
    return 0;
}

File: leetcode1819.go
package main

func max(a, b int) int {
	if a > b {
		return a
	} else {
		return b
	}
}
func gcdd(a, b int) int {
	if b > a {
		a, b = b, a
	}
	if b == 0 {
		return a
	} else {
		return gcdd(b, a%b)
	}
}
func countDifferentSubsequenceGCDs(nums []int) int {
	maxx := 0
	ans := 0
	for _, num := range nums {
		maxx = max(num, maxx)

	}
	vis := make([]int, maxx+1)
	for _, num := range nums {
		vis[num] = 1
	}
	for i := 1; i <= maxx; i++ {
		gcd := 0
		for j := i; j <= maxx; j += i {
			if vis[j] == 1 {
				if gcd == 0 {
					gcd = j
				} else {
					gcd = gcdd(gcd, j)
				}
				if gcd == i {
					ans++
					break
				}
			}
		}
	}
	return ans
}
func main() {
	val := []int{6, 10, 3}
	println(countDifferentSubsequenceGCDs(val))
}
